### PR TITLE
Do not call `initialize_block` before any runtime api

### PR DESCRIPTION
**Pull request by bkchr**
Before this change we always called `initialize_block` before calling
into the runtime. There was already support with `skip_initialize` to skip
the initialization. Almost no runtime_api requires that
`initialize_block` is called before. Actually this only leads to higher
execution times most of the time, because all runtime modules are
initialized and this is especially expensive when the block contained a
runtime upgrade.

TLDR: Do not call `initialize_block` before calling a runtime api.

Fixes: https://github.com/paritytech/polkadot/issues/3053

polkadot companion: https://github.com/paritytech/polkadot/pull/3140

### CODE REVIEW

File: client/light/src/call_executor.rs [hunk 1]
```diff
@@ -230,62 +211,41 @@ pub fn check_execution_proof<Header, E, H>(
 		H: Hasher,
 		H::Out: Ord + codec::Codec + 'static,
 {
-	check_execution_proof_with_make_header::<Header, E, H, _>(
+	check_execution_proof_with_make_header::<Header, E, H>(
```
Comment (tomusdrw) [inline client/light/src/call_executor.rs:214]: Doesn't this break compatibility between proof formats? Are there no consequences?

**Review by tomusdrw — COMMENTED** [review]
Looks good, but I'm surprised there is no docs being updated. I think it's worth to document the behavior now.

File: client/light/src/call_executor.rs [hunk 2]
```diff
@@ -193,30 +185,19 @@ pub fn prove_execution<Block, S, E>(
...
-		&mut changes,
-		"Core_initialize_block",
-		&header.encode(),
-	)?;
-
 	// execute method + record execution proof
 	let (result, exec_proof) = executor.prove_at_trie_state(
 		&trie_state,
-		&mut changes,
+		&mut Default::default(),
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/light/src/call_executor.rs:191]: It doesn't change the behavior to what we had before.

[the remark this answers is not in the document]
Comment (tomusdrw) [reply client/light/src/call_executor.rs:191]: So the `init_proof` that we merged together with `exec_proof` was always empty or redundant?

[the remark this answers is not in the document]
Comment (bkchr) [reply client/light/src/call_executor.rs:191]: That is not what I said here. It was probably not empty as `initialize_block` has done some work. 

However, there are currently no real light clients nor any tests. If we require `initialize_block` to be called before, we probably should add support for passing multiple functions with parameters to the full node.

**Comment by bkchr** [comment]
@pepyakin @tomusdrw burn in etc was all successful. So, now it is time to do a proper review.

File: primitives/consensus/aura/src/lib.rs [hunk 3]
```diff
@@ -90,7 +90,6 @@ sp_api::decl_runtime_apis! {
 		fn slot_duration() -> SlotDuration;
 
 		// Return the current set of authorities.
-		#[skip_initialize_block]
```
[the remark this answers is not in the document]
Comment (bkchr) [reply primitives/consensus/aura/src/lib.rs:93]: This was already prepared by me here: https://github.com/paritytech/substrate/commit/e447c49537e66d0b6e3a408c6ae5c424c7344a7c

From the behavior site also now nothing changes, we did not call `initialize_block` before and now we also don't call it.

File: frame/merkle-mountain-range/primitives/src/lib.rs [hunk 4]
```diff
@@ -406,15 +406,13 @@ sp_api::decl_runtime_apis! {
 	/// API to interact with MMR pallet.
 	pub trait MmrApi<Hash: codec::Codec> {
 		/// Generate MMR proof for a leaf under given index.
-		#[skip_initialize_block]
```
[the remark this answers is not in the document]
Comment (bkchr) [reply frame/merkle-mountain-range/primitives/src/lib.rs:409]: When it did not called `initialize_block` before and now we also don't call `initialize_block`, nothing has changed :shrug:

[the remark this answers is not in the document]
Comment (pepyakin) [reply frame/merkle-mountain-range/primitives/src/lib.rs:409]: it was a brainfart

File: primitives/transaction-pool/src/runtime_api.rs [hunk 5]
```diff
@@ -22,22 +22,32 @@ use sp_runtime::traits::Block as BlockT;
...
+			tx: <Block as BlockT>::Extrinsic,
+		) -> TransactionValidity;
+
 		/// Validate the transaction.
 		///
 		/// This method is invoked by the transaction pool to learn details about given transaction.
 		/// The implementation should make sure to verify the correctness of the transaction
-		/// against current state.
+		/// against current state. The given `block_hash` corresponds to the hash of the block
+		/// that is used as current state.
```
Comment (tomusdrw) [inline primitives/transaction-pool/src/runtime_api.rs:43]: Shall we add info about the need to manually initialize the `frame_system` pallet?

File: primitives/version/src/lib.rs [hunk 6]
```diff
@@ -198,6 +198,11 @@ impl RuntimeVersion {
 	) -> bool {
 		self.apis.iter().any(|(s, v)| s == id && predicate(*v))
 	}
+
+	/// Returns the api version found for api with `id`.
+	pub fn api_version(&self, id: &ApiId) -> Option<u32> {
```
Comment (tomusdrw) [inline primitives/version/src/lib.rs:203]: :pray:

**Review by tomusdrw — APPROVED** [review]
Awesome work!

File: primitives/transaction-pool/src/runtime_api.rs (hunk 5 above)
Comment (bkchr) [reply primitives/transaction-pool/src/runtime_api.rs:43]: For frame this is done automatically through `Executive::validate_transaction`. While this is also just required when you have a signed extension that checks for block hashes.

**Comment by pepyakin** [comment]
Prior this patch, when an runtime API is called we would call `initialize_block`. In turn, it would call `on_initialize` of all modules. `pallet_session` is of special interest, since it would invoke `SessionHandler::on_new_session` as part of its `on_initialize`. With this patch, these routines would not be execution.

Specifically, a difference in behavior will be observed if an runtime API accesses some state that can be changed either in `on_initialize` or `on_new_session`. 

I went through all the runtime API impls and have marked some that do that. Getting into the list doesn't mean that they are wrong, but I think it would be a good idea to check for somebody deeply familiar with the respective code, or at least acknowledge that they are aware of this change.

Here is the list:

- [ ] `BabeApi` @andresilva . Runtime APIs access state which is tainted by `do_initialize` and `enact_epoch_change`. 
**This won't be a big concern**, as far as I can tell. Runtime APIs are called with no digests and thus `do_initialize` does nothing (except randomness collection), `should_epoch_change` returns `false` and thus no `enact_epoch_change`.
- [ ] Equivocation proof generation and verification. AFAICT, this should be fine since we always generate a proper proof. That enables it to be verified in any case.
- [ ] `Beefy` @tomusdrw @adoerr . This one seems can change authorities on new session and thus the `validator_set` API will return a different set at the last block. As I mentioned for BABE it may be fine since it appropriates `ShouldEndSession`.
- [ ] `AuthorityDiscoveryApi` @ordian . On new session this module rotates the keys.

**Review by pepyakin — APPROVED** [review]
Regarding the patch itself, LGTM!

**Comment by tomusdrw** [comment]
Thanks @pepyakin!
I'm crossing-out BEEFY from that list. We are planning to rely on the digest item anyway, currently afair we indeed do the state call and it might be an issue if we have BEEFY validators running two different versions (the ones running older version will think we already transitioned to a new set), but given it's currently deployed only on Rococo it should be easy to coordinate the upgrade, so I don't think any preventive measures are needed.

**Comment by bkchr** [comment]
I checked again Babe and AuthorityDiscovery and couldn't find anything problematic.

**Comment by bkchr** [comment]
bot merge

**Comment by andresilva** [comment]
Both BABE and GRANDPA should be unaffected, I wasn't sure about the equivocation report submission but I tested it now.

**Comment by JoshOrndorff** [comment]
What is the recommended way to port runtime APIs that actually did depend on the initialization? Should each one manually initialize only what it needs to?

cc @tgmichel I believe this is the change that broke our EVM tracing.

**Comment by shawntabrizi** [comment]
@JoshOrndorff not super familiar with how one might initialize the block, but it may be that rather than trying to turn this back on, you actually want to see if you can adjust your code or tracing to work without needing this initialize.

I think in general, it is risky to your chain's performance if you have any of the runtime apis call `initialize_block` because some blocks can be very heavy, and it seems your chain may be attackable on those blocks

**Comment by bkchr** [comment]
Yeah, I would not advise to do this.

If possible, I would prevent this. Or if you only need to initialize some stuff like `System`, you can only initialize this.

**Comment by crystalin** [comment]
@shawntabrizi Thanks for the feedback. 
Just a bit of context, we understand those are a concern for resources (just the fact of fully tracing blocks is very expensive), and that is why we don't expose those features by default. Those are meant for projects to use internally to query the trace logs.
